### PR TITLE
Allocate C arrays for C array literals in an appropriate place

### PR DESCRIPTION
**Pull request by scoder**
* constant C arrays: `static const` C arrays in place
* module global: static C arrays in the module init function
* inside generators: non-const arrays stored in the generator closure
* otherwise: non-const local temp array variables on the stack

The constant case specifically applies to for-loop iteration over strings etc. It avoids the need to store them away in generator closures and generally wasting runtime work for them.

Factors a dedicated `CArrayNode` out of the `ListNode` for the "list as C array" case, to avoid complicating `ListNode` for this. I kept `ListNode` as a base class mostly just in case we need some of its functionality later on, but I'm not aware of anything specific that is needed. It could probably just be a `SequenceNode`. We should be looking at `is_sequence_constructor` rather than specific `ListNode` / `TupleNode` classes wherever it's needed.

Also enables coercion from a Unicode string literal to an integer C array of character values (not just `Py_UCS4[]` but also e.g. `short[]` for more compact storage of BMP Unicode characters). I think that's a reasonable thing to have. It doesn't do any compile time value checking, but the C compiler would probably warn.

It'd be a somewhat large change for 3.2.x but maybe not too risky. Thus also the `ListNode` base class, to reduce the risk of unanticipated changes.

Closes https://github.com/cython/cython/issues/7323

### CODE REVIEW

**Comment by da-woods** [comment]
This doesn't currently seem to work.

I wonder if 3.2.2 should probably have the "safe" version which just disables it, even if this change goes into a later 3.2.x version.

**Comment by scoder** [comment]
> This doesn't currently seem to work.

I think it's fixable, but it might end up as a larger and more risky change in the end.

> I wonder if 3.2.2 should probably have the "safe" version which just disables it, even if this change goes into a later 3.2.x version.

Yeah, we shouldn't wait forever with the next point release.

**Comment by scoder** [comment]
I suspect C++ might not be happy with the `const` C arrays referenced by non-const temp pointers. I'll see how to fix that once I know that this works in general.

**Comment by scoder** [comment]
I was uncertain enough about the testing state of this that I wrote a code generator to build a test suite with everything that I could come up with: for-loops, generators, comprehensions; integers, bytes and unicode strings; etc. Currently cutting a foot path through the issues that those are uncovering as well as the missing cases that I would have expected to be covered.

File: tests/run/carray_iteration.py [hunk 1]
```diff
@@ -0,0 +1,12250 @@
...
+    >>> test_carray_listcomp_const_char_100()
+    [120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120]
+    """
+    carray: cython.pointer[cython.const[cython.char]] = b'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
+
+    return [item for item in carray[:100]]  # cython.const[cython.char]
+
+
+
+def test_carray_setcomp_con
```
Comment (da-woods) [inline tests/run/carray_iteration.py:8614]: This doesn't seem like it will give the right result in the docstring

**Comment by da-woods** [comment]
The Windows ARM failures look like https://developercommunity.visualstudio.com/t/CC-Bad-code-generation-with-large-n/10952684 - we hit it with large calls to `PyTuple_Pack` and `__Pyx_PyList_Pack`.

I was planning to fix it eventually with #7258, but for now I just excluded the two other instances of that bug in windows_arm_bugs.txt

**Comment by da-woods** [comment]
One thing that I found while looking at this:

```
# global scope
cdef const int[3] aaa = [1,2,3] 
```

gives the errors

```c
error: expected expression before ‘[’ token
static int const __pyx_v_7ucoerce_aaa[3] = [1, 2, 3];

warning: passing argument 1 of ‘memcpy’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
 2293 |   memcpy(&(__pyx_v_7ucoerce_aaa[0]), __pyx_carray_, sizeof(__pyx_v_7ucoerce_aaa[0]) * (3));
```

Admittedly it *also* gives the same issues on the master branch though, so wouldn't be a regression here. So I think this could definitely be ignored until later if it made sense.

-----

At function scope Cython just rejects it.

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -9265,6 +9267,97 @@ def generate_operation_code(self, code):
...
+        if array_type.base_type.is_const:
+            array_type = PyrexTypes.c_array_type(
+                PyrexTypes.remove_cv_ref(array_type.base_type), array_type.size)
+        return array_type
+
+    def _allocate_carray(self, code, as_const):
+        array_type = self.type
+        if as_const:
+            # Declare static const array in place.
+            self.array_cname = code.globalstate.new_const_cname('carray')
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:9291]: I don't think it does any harm, but all the examples I can find are allocated in a local scope (either the module-init function, or in an individual function, or in the closure struct) so the names don't need to be globally unique.

Although maybe it makes them easier to reason about if they are.

File: Cython/Compiler/ExprNodes.py [hunk 3]
```diff
@@ -9265,6 +9267,97 @@ def generate_operation_code(self, code):
...
+            # Allocate in the generator closure, cannot be const.
+            array_type = self._non_const_array_type()
+            self.array_cname = code.funcstate.closure_temps.allocate_carray(array_type, self.pos)
+        else:
+            # To be valid C++, we must allocate the memory on the stack
+            # manually and be sure not to reuse it for something else.
+            # Yes, this means that we leak a temp array variable.
+            array_type = self._non_const_array_type()
+            self.array_cname = code.funcstate.allocate_temp(
+                array_type, manage_ref=False, reusable=False)
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:9307]: Why can't you re-use it?

I agree with the "you must allocate it on the stack" (I think there is a nicer way to do it in C), but you are just manually filling in the contents here and you can do that multiple times.

If I set `reusable=True` then it appears to work in all your tests so if you've found a corner case that I've missed (quite likely!) then it isn't explicitly tested.

File: Cython/Compiler/ExprNodes.py (hunk 2 above)
Comment (scoder) [reply Cython/Compiler/ExprNodes.py:9291]: Right, it's not required but I was too lazy to come up with locally unique names and just took a globally unique one instead.

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (scoder) [reply Cython/Compiler/ExprNodes.py:9307]: I don't think it hurt's to keep them separate and it might help the C compiler to see the usage scope. But that's just free guessing. Unless we find a solid reason to make this special, it shouldn't be. I'll take that out.

Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:9307]: It looks like it is actually required; just not in the new tests. Sorry for the false alarm

**Comment by scoder** [comment]
So, then, CI is happy – is this something we can risk putting into 3.2.2?
If so, I'd certainly put out a nightly and an rc first, since it's a rather large change.

**Comment by da-woods** [comment]
How hard is it to put the "simple" fix into 3.2.2 and then this into 3.2.3?

I think https://github.com/cython/cython/issues/7323 is the only significant regression remaining for 3.2 so it might be nice to have a "good" 3.2 release that people can pin to before the risky change. Although possibly this might be more trouble than it's worth.

Not sure... I suspect this is fairly safe as far as large changes go.

**Comment by scoder** [comment]
> put the "simple" fix into 3.2.2 and then this into 3.2.3

Let's do that. The feature is only disabled for generators, which should have a very low impact (famous last words). I'll merge this into the master branch, the other PR into 3.2.x, release 3.2.2 with it, back it out from the maintenance branch and then pick this PR over once we see if it works out.

**Comment by scoder** [comment]
> One thing that I found while looking at this:
> 
> ```
> # global scope
> cdef const int[3] aaa = [1,2,3] 
> ```
> 
> gives the errors
> 
> ```c
> error: expected expression before ‘[’ token
> static int const __pyx_v_7ucoerce_aaa[3] = [1, 2, 3];
> 
> warning: passing argument 1 of ‘memcpy’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
>  2293 |   memcpy(&(__pyx_v_7ucoerce_aaa[0]), __pyx_carray_, sizeof(__pyx_v_7ucoerce_aaa[0]) * (3));
> ```
> 
> Admittedly it _also_ gives the same issues on the master branch though, so wouldn't be a regression here.

This is probably covered by https://github.com/cython/cython/pull/7359
